### ISSUE TITLE
Choose between "test_start_time" vs "creation_time" to compute the age of a test

### ISSUE THREAD

**Issue by pnax** [zonemaster/zonemaster-backend#840]
Currently SQLite uses `test_start_time` (see [SQLite.pm (develop)](https://github.com/zonemaster/zonemaster-backend/blob/develop/lib/Zonemaster/Backend/DB/SQLite.pm#L185..L192)):

```perl
# Search for recent test result with the test same parameters, where "$seconds"
# gives the time limit for how old test result that is accepted.
my ( $recent_hash_id ) = $dbh->selectrow_array(
    "SELECT hash_id FROM test_results WHERE params_deterministic_hash = ? AND test_start_time > DATETIME('now', ?)",
    undef,
    $fingerprint,
    "-$seconds seconds"
);
```

But MySQL and PostgreSQL use `creation_time`:
* [MySQL.pm (develop)](https://github.com/zonemaster/zonemaster-backend/blob/develop/lib/Zonemaster/Backend/DB/MySQL.pm#L171..L176):
    ```perl
    my ( $recent_hash_id ) = $dbh->selectrow_array(
        q[
        SELECT hash_id FROM test_results WHERE params_deterministic_hash = ? AND (TO_SECONDS(NOW()) - TO_SECONDS(creation_time)) < ?
        ],
        undef, $fingerprint, $seconds_between_tests_with_same_params,
    );
    ```

* [PostgreSQL.pm (develop)](https://github.com/zonemaster/zonemaster-backend/blob/develop/lib/Zonemaster/Backend/DB/PostgreSQL.pm#L178..L185):
    ```perl
    my $sth = $dbh->prepare( "
        INSERT INTO test_results (batch_id, priority, queue, params_deterministic_hash, params, undelegated)
        SELECT ?, ?, ?, ?, ?, ?
        WHERE NOT EXISTS (
            SELECT * FROM test_results
            WHERE params_deterministic_hash = ?
              AND creation_time > NOW() - ?::interval
    )" );
    ```

It would be nice to use the same value for all database engines.

Now which one should we choose?

**Comment by pnax** [comment]
> Now which one should we choose?

The [current API documentation (develop)](https://github.com/zonemaster/zonemaster-backend/blob/develop/docs/API.md#api-method-start_domain_test) says:

```
If an identical test was already enqueued and hasn't been started or
was enqueued less than 10 minutes earlier, no new test is enqueued. 
```

And in the [Configuration file (develop)](https://github.com/zonemaster/zonemaster-backend/blob/develop/docs/Configuration.md#age_reuse_previous_test):

```
If a new test is requested for the same zone name and parameters
within the shelf life of a previous test result, that test result is reused.
Otherwise a new test request is enqueued.
```

So I would say that we should update the SQLite logic to use the `creation_time` value.

**Comment by matsduf** [comment]
In a way I think `creation_time` makes most sense, but we should look at this in the light of batches. If we let tests created in a batch get the `creation_time` when it is created and `test_start_time` then tests in a batch waiting to be tested has 0 or null in `test_start_time` and then if a single test comes for a domain in the batch queue we will not get the unwanted behavior that a non-started test is returned.

Today `test_start_time` is updated after start, at least by some DB engine, and that is probably not what we want.

**Comment by matsduf** [comment]
> > Today `test_start_time is` updated after start, at least by some DB engine, and that is probably not what we want.
> 
> Well, to me this is what we want.

I meant that it is first set and then updated to an new time. I think it should be set once and that is when the test starts.

> I think `test_start_time` should be unset until the test is started, which is when it is taken by the TestAgent and that its `progress` is set to `1`. Looking at the current code, it seems that SQLite and MySQL set the `test_start_time` on test creation **and** when the test is started by the TestAgent. I really think we should avoid touching the `test_start_time` before beginning the test. (see #842)

Agree.

> Note that `test_start_time` is reset when collecting unfinished tests. This collect does use the `test_start_time` value and look if it is greater than the [`ZONEMASTER.max_execution_time`](https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#max_zonemaster_execution_time) (see `schedule_for_retry()`). Currently this reset `test_start_time` to the current time, I think this should be reset to `null`.

If unfinished tests are just set to done I see no reason to change the `test_start_time`, but I guess if tests are to be restarted setting it to `null` seems reasonable.

> Now, about the `creation_time` and batch job, what about updating the test's `priority`? Since the problem we encounter is when the test was added by a batch job (usually with a low priority) and someone likes to test this same domain, we can increase the test's `priority` (for instance to to the one of a single test). Hence the test will be caught and run by the TestAgent sooner (see `get_test_request()`). And we don't create a new test, since `get_test_request()` returns the test created by the batch job.

I think it is better not to depend on `priority`. I see no problem that a second test is started. There is no problem as such that the same domain can be tested at the same time, one single test and one from a batch. That could also happen if the same domain is started with different DS values or in different batches.

**Comment by pnax** [comment]
> If unfinished tests are just set to done I see no reason to change the test_start_time, but I guess if tests are to be restarted setting it to null seems reasonable.

Currently there is no way to know the _state_ of a test. This might be something to consider in future work. When we have unfinished tests they are reschedule, and this implies resetting its `progress` to `0` and I suggest to reset `test_start_time` to `NULL` as well.

> I think it is better not to depend on `priority`. I see no problem that a second test is started.

Since the search for a similar test is based on the fingerprint, if we agree to no depend on the `priority`, what if we re-add this value to the computation of the `fingerprint`? (this could be done in #831)

**Comment by matsduf** [comment]
> > I think it is better not to depend on `priority`. I see no problem that a second test is started.
> 
> Since the search for a similar test is based on the fingerprint, if we agree to no depend on the `priority`, what if we re-add this value to the computation of the `fingerprint`? (this could be done in #831)

I do not see the meaning of including priority. Fingerprint should give tests that probably give the same result, and priority has no affect on that.

**Comment by pnax** [comment]
> Fingerprint should give tests that probably give the same result, and priority has no affect on that.

Good point.

So if a test is first created by a batch job and a similar test is created by a user with the same parameters (same fingerprint), we should always create a new test and return its id.

**Comment by mattias-p** [comment]
> Currently there is no way to know the state of a test. This might be something to consider in future work.

Yes, definitely. I would be really helpful to have a state field in the tests table. The life cycle of a test should also be documented in terms of a finite state machine.

Regarding reuse, what I'm hearing is that batches should never reuse any pre-existing tests, and that batch tests should never be considered for reuse. Is that correct?

**Comment by pnax** [comment]
> Regarding reuse, what I'm hearing is that batches should never reuse any pre-existing tests, and that batch tests should never be considered for reuse. Is that correct?

I would say that so far we have only discussed the second part (ie: "batch tests should never be considered for reuse"). Regarding the fact that batches could reuse a test, this has not been discussed as far as I know. Thinking about this, I can't see anything preventing a test reuse for batches. I would even argue that this will be a benefit since it will reduce the amount of tests for the batch.

To summarize (with regard to the https://github.com/zonemaster/zonemaster-backend/issues/840#issuecomment-895522399 suggesting not to update the `priority`) we should not reuse batch tests. But it's okay for batches to reuse pre-existing tests.
If we agree on this, I can implement such logic and add it to #831.

**Comment by pnax** [comment]
> But it's okay for batches to reuse pre-existing tests.

And now I realize that the current implementation would need to update the `batch_id` value of the pre-existing test. If we choose not to update the `priority` of a pre-existing test, I can't see why we should update the `batch_id` column either. Therefore it looks like what you were correct @mattias-p, we should never consider batch tests for reuse (if we decide not to update pre-existing tests).
This is tricky.

Maybe we first need to better review our needs here.

I think we should choose one of the following solutions (with "batch test" a test created by a batch job and "single test" a test not created by a batch job):

1. We have two disjoints sets: one with only "batch tests" and one with "single tests"
2. It is possible to reuse pre-existing "single test" for batches
3. It is possible to reuse pre-existing "batch test" for "single test"
4. We allow both 2. and 3.

Any thoughts on this?

I think that if we decide to properly reuse tests, we might need the state field discuss earlier. So I would suggest to start with the easiest, use two disjoint sets.

**Comment by mattias-p** [comment]
Another question is whether to reuse tests between batches.

If a test is requested and we have an test with equivalent parameters that's still good (i.e. it hasn't expired), is it ever a bad thing to reuse that test? The only difference I see is faster response times for the client and lower resource usage for the server. I don't think implementation complexity is an issue. Maximizing the reuse would require a schema update but we're doing those occasionally anyway. And the conceptual complexity would be lower.

**Comment by pnax** [comment]
> Another question is whether to reuse tests between batches.

When I talked about disjoint sets, I had in mind that a test can be reused within its set (but not between the sets). We look for a pre-existing test within each respective set.

I admit that my last comment was not very clear. Here is an update:

1. We have two disjoints sets: one with "batch tests" and one with "single tests". A test can only be reused within its set.
2. We have two sets: one with "batch tests" and one with "single tests". It is possible to reuse a "single test" for batches.
3. We have two sets: one with "batch tests" and one with "single tests". It is possible to reuse a "batch test" for "single test".
4. We have only one set made of "batch tests" and "single tests". Any test can be reused.

I also share your view on reusing tests.

**Comment by matsduf** [comment]
I agree, option 4 is the most reasonable. In practice, however, there will be few instances when the same domain name (with the same setting) is tested both in a batch and as a single test within the short time frame set for reuse of result.

**Comment by matsduf** [comment]
@pnax, is this resolved?

**Comment by pnax** [comment]
Well no, I've just found that commit 5df15c0e402cb049773bfdc45adcd8747c42fe85 removed the update made in a8a4cee3f37cef0a3a62574c1ecd18219a26788b. So we are back with an inconsistency.

**Comment by pnax** [comment]
#769 introduced the following change which settles the discussion.

https://github.com/zonemaster/zonemaster-backend/blob/9f733dd58d925efc660ee6486e1ab6f815911fdf/lib/Zonemaster/Backend/DB.pm#L169-L170
